completions/bash/poudriere

```diff
@@ -0,0 +1,437 @@
+#    notice, this list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+# SUCH DAMAGE.
+#
+# ---
+# Version: 0.1.0
+
+_poudriere_absolute_path() {
+    COMPREPLY=($(compgen -f "${cur:-/}" -- "$cur"))
+}
+
+# XXX: It might be non exhaustive.
+_poudriere_architecture() {
+    COMPREPLY=($(compgen -W 'amd64 i386' -- "$cur"))
+}
+
+# TODO: Parse poudriere status.
+_poudriere_buildname() {
+    compopt -o default
+}
+
+_poudriere_direct_port() {
+    # Find a ports tree to search for ports.
+    local prev_word=
+    local ports_tree=
+    # Find out if a ports tree is already specified with -p.
+    for w in ${words[@]}; do
+        if [[ $w == -p ]]; then
+            prev_word=-p
+        elif [[ $prev_word == -p ]]; then
+            ports_tree="$w"
+            break
+        fi
+    done
+    # Find out if the default tree is present.
+    if [[ -z $ports_tree ]] && poudriere ports -l | awk '/^default */{found=1} END{if (found != 1) {exit 1}}'; then
+        ports_tree=default
+    elif [[ -z $ports_tree ]]; then
+        return
+    fi
+    local ports_tree_path="$(poudriere ports -l | awk -v tree=$ports_tree 'NR > 1{if (match($1, tree)) {print $5}}')"
+    # Complete port name.
+    if [[ $cur =~ .*/.* ]]; then
+        COMPREPLY=($(compgen -W "$(
+        command cd "$ports_tree_path" && LANG=C find "${cur%%/*}/" -mindepth 1 -maxdepth 1 -type d -name "${cur##*/}*"
+        )" -- "$cur"))
+    else # Complete category.
+        COMPREPLY=($(compgen -W "$(command cd "$ports_tree_path" && LANG=C find * -maxdepth 0 -type d -name '[^.A-Z]*' -prune)" -S "/" -- "$cur")) && compopt -o nospace
+    fi
+}
+
+_poudriere_filesystem() {
+    COMPREPLY=($(compgen -W "none $(zfs list | awk 'NR > 1{print $1}')" -- "$cur"))
+    if [[ ${#COMPREPLY[@]} == 1 ]] && [[ $COMPREPLY != none ]]; then
+        # XXX: Is it reasonable to assume that someone would continue to append somthing to an existing name?
+        compopt -o nospace
+    fi
+}
+
+_poudriere_freebsd_version() {
```

Comment:
This is going to become bitrotted fast.

Reply:
I've found a way to generate it dynamically :) Stay tuned.